### DISCUSSION TITLE
Be able to import only VTK classes that do not need OpenGL

### DISCUSSION THREAD

**Discussion by toro01** [pyvista/pyvista#3283] [Ideas]
Hi all,

I'm using PyVista for its ability to post process VTK data structures in a very pythonic and concise way. It works great.

But, I realize that, even if I don't plot anything, VTK classes relying on OpenGL are loaded as soon as I do `import pyvista as pv`, or even `from pyvista import Sphere, RectilinearGrid`. The same functionality written with VTK Python doesn't requires OpenGL to be installed.

In environment where OpenGL is missing, I typically get this exception:

>     import pyvista as pv
> /github/home/.local/lib/python3.7/site-packages/pyvista/__init__.py:12: in <module>
>     from pyvista.plotting import *
> /github/home/.local/lib/python3.7/site-packages/pyvista/plotting/__init__.py:4: in <module>
>     from .charts import Chart2D, ChartMPL, ChartBox, ChartPie
> /github/home/.local/lib/python3.7/site-packages/pyvista/plotting/charts.py:13: in <module>
>     from pyvista import _vtk
> /github/home/.local/lib/python3.7/site-packages/pyvista/_vtk.py:444: in <module>
>     from vtkmodules.vtkRenderingOpenGL2 import (
> E   ImportError: libGL.so.1: cannot open shared object file: No such file or directory

(whatever the python version is)

This is a problem for 2 reasons:
1. OpenGL is not needed in my use case. This useless dependency might have consequences in terms of security, maintainability, etc...
2. OpenGL must be installed in the CI (as a workaround, I found https://github.com/pyvista/setup-headless-display-action but it doesn't work in a container as sudo is not installed) and in the customers's environments. As they may use the project on headless machines, they will have to install opengl (not installed by default), Xvbf or any other alternative, launch a script to initialize everything, etc... So lots of useless boilerplate.

My suggestion: would it be possible to load only VTK modules that do not depend on OpenGL when doing import `pyvista as pv`? Or at least when doing specialized imports like `from pyvista import Sphere, RectilinearGrid`?

Hope this makes sense. Thanks for the good work.

**Comment by akaszynski** [comment]
Lazy imports is something we've been meaning to do for a while. We've operated under the assumption of importing everything as needed in `_vtk.py`, which is sort of our internal version of `vtkmodules` and it makes it easier to track imports. However, this still requires the import of a variety of libraries that may or may not be used and this has several issues (as you pointed out) to include:
- Longer import time of `pyvista`
- Dependencies on potentially unnecessary libraries.

This is something that we can (and really should) clean up, but it's going to be a non-trivial effort. It looks like Python is adding support for this in [PEP 690](https://peps.python.org/pep-0690/), but until this has landed and is supported by the oldest version of Python we support, we're going to have to stick with a third-party library or a home grown lazy importer if we want to implement this.

**Reply by akaszynski** [reply]
Actually, this might not be so bad:
https://docs.python.org/3/library/importlib.html#implementing-lazy-imports

Working...

**Reply by akaszynski** [reply]
After investigating this, I've determined the following:
- If our classes inherit from a VTK class and we import that class at the top level (e.g. `pyvista.<class>`) then we must load in the required VTK libraries, and as a result, OpenGL. Unless we decide to implement a hack (e.g. monkey patch VTK classes) at runtime, there's no getting around this as within Python dynamic inheritance would require a class factory and we can't override the base class.

Proposed workarounds:
- Class Composition. We do this for a few of our classes and I'm starting to think it's the better (albeit time consuming) approach. This way we can dynamically load VTK libraries outside of `_vtk` and instead in the class itself and eventually eliminate or trim down the number of VTK modules exposed within `_vtk`. This would be backwards compatible and would have the advantage of separating out all VTK methods into a consistent attribute (a public attribute like `vtkclass`). Would require a significant refactor.
- Move classes out of the top-level `pyvista`, move most VTK imports into . This would not be backwards compatible unless we implemented a module level `getattr` combined with a lazy import. I can see most code inspectors complaining about this would make it difficult, if not impossible, to inspect our code base).

@pyvista/developers, your thoughts?

**Reply by adeak** [reply]
I haven't thought much about this yet, but as a knee-jerk reaction I'd caution against drastic black magic such as module `__getattr__` for the sole purpose of supporting edge cases like this one (which is not to say the use case is not a valid one, just the price to pay this way too much).

Another note is that switching to composition would also be a breaking change unless we lie that our classes remain instances of the wrapped VTK class. Downstream will probably often assume that `isinstance(mesh, vtk.vtkImageData)` covers `pyvista.UniformGrid` as well.

Can we instead just hack a `try`-`except` guard around the OpenGL imports, warning in case of failure that the package will be hamstrung? Downstream users intentionally missing OpenGL could silence the warning if it's specific enough.

**Reply by akaszynski** [reply]
> Can we instead just hack a try-except guard around the OpenGL imports, warning in case of failure that the package will be hamstrung? Downstream users intentionally missing OpenGL could silence the warning if it's specific enough.

This would work and be the least backwards breaking, but will make a mess of `_vtk` and the `plotting` module. This might be a good enough solution. I'd have to test this out on a fresh VM, but this seems to me the easiest to implement and we guard only those needing opengl.

Any other dependencies while we're at it?

**Reply by adeak** [reply]
What I have in mind doesn't leave `_vtk`. For the few select dependencies (OpenGL for starters), warn if import fails and assign `None` to all the names that would otherwise be imported. If users go ahead and try to use functionality that is missing, we will crash. "Best effort" attempt but all bets are off.

Looking at `_vtk.py` there don't seem to be too many OpenGL-y imports, so the result shouldn't be too messy.

**Reply by akaszynski** [reply]
I don't think assigning None would work:
```py
try:
    from vtkmodules.vtkChartsCore import vtkChartXY
    raise ImportError
except ImportError:
    vtkChartXY = None
    

class Chart(vtkChartXY):
    pass
```

```
~/tmp/tmp.py in <module>
      6 
      7 
----> 8 class Chart(vtkChartXY):
      9     pass

TypeError: NoneType takes no arguments
```

However, `object` works.
```py
try:
    from vtkmodules.vtkChartsCore import vtkChartXY
    raise ImportError
except ImportError:
    vtkChartXY = object
    

class Chart(vtkChartXY):
    pass
```

**Reply by adeak** [reply]
Ah, from a quick glance I didn't think they got subclassed.

If we're going with something like this, I wonder if it might be clearer then to use an alias, like
```py
MISSING_DEPENDENCY = object

...

try:
    from vtkmodules.vtkChartsCore import vtkChartXY
    raise ImportError
except ImportError:
    vtkChartXY = MISSING_DEPENDENCY
```
Or upping the ante a bit and using an abstract class that raises when instantiated (not sure if this can be done without touching the library code, i.e. without metaclasses). Subclassing `object` might tick some boxes, postponing errors (until the first inherited method access is attempted).

**Reply by adeak** [reply]
Promising:
```py
>>> class ScreamingABC:
...     def __new__(*args):
...         raise TypeError(f'Instantiation of {args[0]} is not possible!')
... class Foo(ScreamingABC):
...     pass
... Foo()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [28], in <cell line: 6>()
      4 class Foo(ScreamingABC):
      5     pass
----> 6 Foo()

Input In [28], in ScreamingABC.__new__(*args)
      2 def __new__(*args):
----> 3     raise TypeError(f'Instantiation of {args[0]} is not possible!')

TypeError: Instantiation of <class '__main__.Foo'> is not possible!
```
But this needs the screaming ABC to be the first type to inherit from.

**Reply by banesullivan** [reply]
Could we just split the `_vtk` module up in the different PyVista submodules? If we isolate properly, then I think this is doable.

For instance, in `pyvista/core/` we could have a `_vtk_core` module that has only the VTK depends for our core classes and algorithms. Charts and Plotting would do their own too.

In the `_vtk_core` module, we'd have all the `vtkmodules.vtkCommonDataModel` imports etc

I assume @toro01's goal is to use the core data classes without OpenGL rather than Charts?

**Reply by banesullivan** [reply]
Then we could maybe hook it up so that PyVista's plotting and charts (and any OpenGL dependent) modules are lazily imported (rather than lazily importing VTK).

.... this is all just speculative thoughts as I drive by on all my GitHub notifications though

**Reply by toro01** [reply]
> Could we just split the `_vtk` module up in the different PyVista submodules? If we isolate properly, then I think this is doable.
> 
> For instance, in `pyvista/core/` we could have a `_vtk_core` module that has only the VTK depends for our core classes and algorithms. Charts and Plotting would do their own too.
> 
> In the `_vtk_core` module, we'd have all the `vtkmodules.vtkCommonDataModel` imports etc
> 
> I assume @toro01's goal is to use the core data classes without OpenGL rather than Charts?

Sorry for the delay. Correct, I don't need Charts. Because of this "issue", I'm going with VTK Python for the moment and the imports I need are the following:
```py
from vtkmodules.vtkCommonCore import vtkDoubleArray, vtkFloatArray
from vtkmodules.vtkCommonDataModel import vtkRectilinearGrid
from vtkmodules.vtkFiltersCore import vtkPointDataToCellData, vtkProbeFilter
from vtkmodules.vtkFiltersGeometry import vtkRectilinearGridGeometryFilter
from vtkmodules.vtkFiltersSources import vtkPointSource
from vtkmodules.vtkCommonCore import (
    vtkDoubleArray,
    vtkIntArray,
    vtkPoints,
    vtkStringArray,
)
from vtkmodules.vtkCommonDataModel import (
    VTK_PIXEL,
    VTK_VERTEX,
    VTK_VOXEL,
    vtkRectilinearGrid,
    vtkUnstructuredGrid,
)
from vtkmodules.vtkIOXML import vtkXMLDataSetWriter
```
Thanks.
